dip-0008.md

```diff
@@ -275,6 +277,109 @@ resolve any ambiguities which might occur in the last few blocks.
+binomial coefficients and have a delightful relation to Pascal's Triangle.
+
+For any process with an observable outcome the set of all outcomes is
+called a sample space.  If all outcomes are equally likely, as in the
+selection of a quorum, the probability of any event is the number of
+outcomes that satisfy that event over the total number outcomes.
+
+An attacker can temporarily prevent ChainLocks if they control
+160 out of 400 masternodes in the quorum.  If the attacker controls
+`m` out of the `N` total masternodes we can compute the
+number of outcomes that satisfy the event "exactly 160
+nodes of the quorum are controlled by the masternode."
+Specifically there are
+<sub>`m`</sub>`C`<sub>`160`</sub> * <sub>`N - m`</sub>`C`<sub>240</sub>
+outcomes where the attacker controls exactly `160` masternodes.
+Thus, the probability that the attacker controls exactly `160` nodes is
+
+<p align="center"><img src="dip-0008/ThreshCase.gif" alt="Threshold Probability"></p>
+
+The attacker could also be successful if they control <b>more</b> than `160`
+of the nodes in the quorum selection. Thus the exact probability that an
+attacker can successfully prevent ChainLocks for that quorum round is
+computed by counting the number of outcomes where the attacker controls
+160 or more nodes in the quorum divided by the total number of outcomes.
+This is computed by the sum
+
+<p align="center"><img src="dip-0008/fin_sum.gif" alt="Final sum" ></p> 
+  
+The numbers involved in this sum can be quite large. We may
+use a <A HREF="dip-0008/quorum_attack.py">Python Script</A> to effectively
+evaluate this sum. Note
+that we made the necessary modifications to this formula
+to calculate the probability that an attacker controls 60% or more
+of the quorum.
+
+| N | m |Type of Attack|success probability in one trial |
+|---|---|--------------|--------------------------------|
+|5000|500| withhold |  2.79e-64|
+|5000|1000| withhold | 5.09e-22 |
+|5000|1500| withhold | 5.59e-6 |
+|5000|500| malicious | 7.11e-157 |
+|5000|1000| malicious | 2.86e-76 |
+|5000|1500| malicious | 1.89e-38 |
+|2000|200 | withhold | 0 |
+|2000|400 | withhold | 6.88e26 |
+|2000|600 | withhold | 1.12e-6 |
+|2000|200 | malicious | 0 |
+|2000|400 | malicious | 9.49e-94 |
+|2000|600 | malicious | 3.93e-45 |
+
+In all scenarios above, an attacker should not expect to withhold a
+ChainLock in the next century. The attacker would have a less than
+one in 100 trillion chance of producing at least
+one malicious ChainLock in the next sextillion (10^21) years.
+
+From these calculations we conclude that an attacker with control of fewer
+than 30% of all masternodes will not have any influence on ChainLocks. For
+completeness we note that
+if an attacker controlled 50% of masternodes an attacker would
+generally be able to prevent ChainLocks, but there is less than 30 out of
+one billion chance that the attacker will be able to produce at least one
+malicious ChainLock in the next million years.
```

Comment:
These numbers assume the most optimistic scenario that the quorum is constructed with a full 400 participating nodes (which is actually pretty close to how things are working). The other end of the spectrum is limited by the [`quorumMinSize` parameter](https://github.com/thephez/dips/blob/master/dip-0006.md#current-llmq-types). For the large LLMQ used by ChainLocks, the allowable number of participating members is 300-400. If the quorum only had 300 members, the threshold for withholding a ChainLock is substantially lower.

Reply:
I appreciate these comments and will make edits so that this is clear to the reader.

Reply:
I alerted the reader that this is the case with a new paragraph in 1f622c903aefff6cfc8df6dcab202e74bf59f05a